### PR TITLE
When reporting URL of a link use caret position instead of navigator object

### LINKED ISSUE #14659: Global command for reporting destination of a link: Consider reporting for the link at the position of the virtual caret / focus rather than for the navigator object
### Is your feature request related to a problem? Please describe.
In #14583 @XLTechie  added a global commands which allow to report a URL for the link at the position of the navigator object. While this is a **huge** improvement compared to either having copy the URL to the clipboard or accessing the URL from the status bar from the web browser, the fact that these scripts report destination for the link at the navigator object rather than focus / caret position makes their usage impractical for people who prefer to work with review cursor / navigator object not following focus / caret.

### Describe the solution you'd like
Ideally these new scripts should report the URL for the link which is at the position of the virtual caret in  browse mode or for the object with focus. For most users there should be no difference in behavior, since by default navigator object follows focus.
### Describe alternatives you've considered
Leave the behavior as is.
### Additional context
I'm not sure why the navigator object was chosen in the first place, so perhaps there is a good reason which I'm not aware of. Assuming this proposal is accepted we would also have to decide if we need additional scripts for reporting destination of the link at the position of the navigator object - I can't think about any use cases, but perhaps there are some. cc @michaelDCurran and @Qchristensen  for additional input, since both of you reviewed the original implementation.
If this proposal is accepted I would be happy to submit an implementation.

### PR DESCRIPTION
**Pull request by lukaszgo1**
### Link to issue number:
Closes #14659
### Summary of the issue:
In #14583 @XLTechie added a global commands which allows to report a URL for the link at the position of the navigator object. While this is a huge improvement compared to either having copy the URL to the clipboard or accessing the URL from the status bar from the web browser, the fact that these scripts report destination for the link at the navigator object rather than focus / caret position makes their usage impractical for people who prefer to work with review cursor / navigator object not following focus / caret.
### Description of user facing changes
When user requests URL of a link NVDA reports URL of a link at the position of the caret / focus.
### Description of development approach
Script for reporting URL was modified so that it first tries to get an object at the current caret position, and if that fails from the focused object. User guide and documentation for global commands were updated accordingly.
After @XLTechie's review it was noticed that the script fails to work for links without `href` attribute, as part of this PR I've added handling for such links.
### Testing strategy:
Tested various objects which are links, linked graphics, and non links in both focus and browse mode, in Firefox and IE.
### Known issues with pull request:
None known
### Change log entries:
Changes
- Script for reporting destination of a link now reports destination for the link at the caret / focus position rather than for the navigator object.

Bug fixes:
- When trying to report URL for links without `href` attribute NVDA is no longer silent - the fact that the given link has no destination is reported instead.
### Code Review Checklist:
- [X] Pull Request description:
  - description is up to date
  - change log entries
- [X] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [X] API is compatible with existing add-ons.
- [X] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [X] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [X] Security precautions taken.

### CODE REVIEW

**Comment by lukaszgo1** [comment]
@XLTechie  You may want to review / test this.

**Comment by XLTechie** [comment]
I shall. In all of my trials for this, I could never get it to work reliably 
enough from a user POV, so I'll be interested to see what you've done.

File: source/globalCommands.py [hunk 1]
```diff
@@ -3630,8 +3630,8 @@ def script_reloadPlugins(self, gesture):
 
 	@script(
 		description=_(
-			# Translators: input help mode message for Report destination URL of navigator link command
-			"Report the destination URL of the link in the navigator object. "
+			# Translators: input help mode message for Report destination URL of a link command
+			"Report the destination URL of the link at the current user position. "
```
Comment (XLTechie) [inline source/globalCommands.py:3634]: "user position" is a vague concept.

```suggestion
			"Report the destination URL of the currently focused link or graphic."

```

File: source/globalCommands.py [hunk 2]
```diff
@@ -3640,11 +3640,17 @@ def script_reloadPlugins(self, gesture):
 	def script_reportLinkDestination(
 			self, gesture: inputCore.InputGesture, forceBrowseable: bool = False
 	) -> None:
-		"""Generates a ui.message or ui.browseableMessage of a link's destination, if the navigator
-		object is a link, or an element with an included link such as a graphic.
+		"""Generates a ui.message or ui.browseableMessage of a link's destination, if user is positioned on a link,
```
Comment (XLTechie) [inline source/globalCommands.py:3643]:
```suggestion
		"""Generates a ui.message or ui.browseableMessage of a link's destination, if focus or caret is positioned on a link,

```

File: user_docs/en/userGuide.t2t [hunk 3]
```diff
@@ -222,7 +222,7 @@ The actual commands will not execute while in input help mode.
 | Read status bar | ``NVDA+end`` | ``NVDA+shift+end`` | Reports the Status Bar if NVDA finds one. Pressing twice will spell the information. Pressing three times will copy it to the clipboard |
 | Read time | ``NVDA+f12`` | ``NVDA+f12`` | Pressing once reports the current time, pressing twice reports the date |
 | Report text formatting | ``NVDA+f`` | ``NVDA+f`` | Reports text formatting. Pressing twice shows the information in a window |
-| Report link destination | ``NVDA+k`` | ``NVDA+k`` | Pressing once speaks the destination URL of the link in the [navigator object #ObjectNavigation]. Pressing twice shows it in a window for more careful review |
+| Report link destination | ``NVDA+k`` | ``NVDA+k`` | Pressing once speaks the destination URL of the link at the current position. Pressing twice shows it in a window for more careful review |
```
Comment (XLTechie) [inline user_docs/en/userGuide.t2t:225]:
```suggestion
| Report link destination | ``NVDA+k`` | ``NVDA+k`` | Pressing once speaks the destination URL of the link at the current focus or caret position. Pressing twice shows it in a window for more careful review |

```

**Review by XLTechie — CHANGES_REQUESTED** [review]
I think this will be fine. I'm not really sure if it covers more common use cases, but it appears to be slightly more flexible, at the cost of making the review cursor a bit less useful for link identification with some configurations. That probably won't be a problem for most users, and as stated, most likely won't notice either way.

**Comment by XLTechie** [comment]
@lukaszgo1 As an aside, you may want to handle the case of the very first link/graphic (Ctrl+Home position) on the Appveyor website. Noticed when downloading the try build for this PR in Firefox.

"Not a link" is not announced, yet there is no link.

Perhaps if Value is None, we should also announce "not a link"? I would like to say "Link has no apparent destination", because it might still be clickable somehow.

If you try to open the Window version of reportLinkDestination, this happens:

```
IO - inputCore.InputManager.executeGesture (04:47:09.187) - winInputHook (20220):
Input: kb(desktop):NVDA+k
IO - inputCore.InputManager.executeGesture (04:47:09.325) - winInputHook (20220):
Input: kb(desktop):NVDA+k
ERROR - scriptHandler.executeScript (04:47:09.340) - MainThread (19360):
error executing script: <bound method GlobalCommands.script_reportLinkDestination of <globalCommands.GlobalCommands object at 0x0719EB90>> with gesture 'NVDA+k'
Traceback (most recent call last):
  File "scriptHandler.pyc", line 289, in executeScript
  File "globalCommands.pyc", line 3659, in script_reportLinkDestination
  File "ui.pyc", line 106, in browseableMessage
  File "html\__init__.pyc", line 19, in escape
AttributeError: 'NoneType' object has no attribute 'replace'
```

If you just try the speaking report, you get the following (dev info included):

```
IO - speech.speech.speak (03:30:59.874) - MainThread (2428):
Speaking ['heading', 'level 1', 'link', 'AppVeyor']
IO - inputCore.InputManager.executeGesture (03:31:01.816) - winInputHook (18724):
Input: kb(desktop):NVDA+k
IO - inputCore.InputManager.executeGesture (03:31:03.248) - winInputHook (18724):
Input: kb(desktop):NVDA+f1
DEBUG - gui.contextHelp.bindHelpEvent (03:31:03.279) - MainThread (2428):
Did context help binding for LogViewer
INFO - globalCommands.script_navigatorObject_devInfo (03:31:03.334) - MainThread (2428):
Developer info for navigator object:
name: 'AppVeyor'
role: Role.LINK
processID: 11596
roleText: None
states: State.LINKED
isFocusable: False
hasFocus: False
Python object: <NVDAObjects.IAccessible.mozilla.Mozilla object at 0x04B1A970>
Python class mro: (<class 'NVDAObjects.IAccessible.mozilla.Mozilla'>, <class 'NVDAObjects.IAccessible.ia2Web.Ia2Web'>, <class 'NVDAObjects.IAccessible.IAccessible'>, <class 'NVDAObjects.window.Window'>, <class 'NVDAObjects.NVDAObject'>, <class 'documentBase.TextContainerObject'>, <class 'baseObject.ScriptableObject'>, <class 'baseObject.AutoPropertyObject'>, <class 'garbageHandler.TrackedObject'>, <class 'object'>)
description: ''
location: RectLTWH(left=370, top=113, width=144, height=54)
value: None
TextInfo: <class 'NVDAObjects.IAccessible.IA2TextTextInfo'>
appModule: AppModule(appModuleHandler, appName='firefox', processID=11596)
appModule.productName: 'Firefox'
appModule.productVersion: '111.0'
appModule.helperLocalBindingHandle: c_long(116676168)
windowHandle: 2426720
windowClassName: 'MozillaWindowClass'
windowControlID: 0
windowStyle: 382664704
extendedWindowStyle: 256
windowThreadID: 11600
windowText: 'nvda pr14723-27847,9913f99d - AppVeyor — Mozilla Firefox'
displayText: ''
IAccessibleObject: <POINTER(IAccessible2) ptr=0x6fa349c at 8ee030>
IAccessibleChildID: 0
IAccessible event parameters: windowHandle=2426720, objectID=-4, childID=-159383565
IAccessible accName: 'AppVeyor'
IAccessible accRole: ROLE_SYSTEM_LINK
IAccessible accState: STATE_SYSTEM_LINKED, STATE_SYSTEM_VALID (4194304)
IAccessible accDescription: ''
IAccessible accValue: None
IAccessible2 windowHandle: 2426720
IAccessible2 uniqueID: -159383565
IAccessible2 role: ROLE_SYSTEM_LINK
IAccessible2 states: IA2_STATE_OPAQUE, IA2_STATE_SELECTABLE_TEXT (5120)
IAccessible2 attributes: 'display:block;class:collapse-only;margin-bottom:0px;tag:a;text-indent:-9999px;formatting:block;text-align:left;margin-right:0px;margin-left:0px;margin-top:0px;'
IAccessible2 relations: containingDocument * 1, containingTabPane * 1, containingApplication * 1
```

File: source/globalCommands.py (hunk 1 above)
Comment (lukaszgo1) [reply source/globalCommands.py]: I've rephrased the description slightly, to consistently mention caret / focus everywhere these commands are documented.

**Comment by lukaszgo1** [comment]
> [...] I'm not really sure if it covers more common use cases, but it appears to be slightly more flexible, at the cost of making the review cursor a bit less useful for link identification with some configurations. 

The only hypothetical case I can think of would be a link which is not reachable in browse mode, and is not focusable. I haven't seen such a control anywhere, so I believe this case can be safely ignored.
> As an aside, you may want to handle the case of the very first link/graphic (Ctrl+Home position) on the Appveyor website. Noticed when downloading the try build for this PR in Firefox.
"Not a link" is not announced, yet there is no link.
Perhaps if Value is None, we should also announce "not a link"? I would like to say "Link has no apparent destination", because it might still be clickable somehow.

I've implemented your suggestion, since apparently links with no `href` are valid in HTML.

File: source/globalCommands.py [hunk 4]
```diff
@@ -3667,8 +3681,8 @@ def script_reportLinkDestination(
 
 	@script(
 		description=_(
-			# Translators: input help mode message for Report URL of navigator link in a window command
-			"Reports the destination URL of the link in the navigator object in a window, "
+			# Translators: input help mode message for Report URL of a link in a window command
+			"shows the destination URL of the link at the position of caret or focus in a window, "
```
Comment (seanbudd) [inline source/globalCommands.py:3685]: Displays might be more accurate? Also fixing the capital letter for the start of the sentence

```suggestion
			"Displays the destination URL of the link at the position of caret or focus in a window, "
```

```suggestion
			"Shows the destination URL of the link at the position of caret or focus in a window, "
```

**Review by seanbudd — APPROVED** [review]
Generally LGTM

**Review by Qchristensen — CHANGES_REQUESTED** [review]
Would the user guide wording be more correct as "speaks the destination URL of the link at the caret position or current focus" (as it tries the caret position first?)

**Review by seanbudd — APPROVED** [review]

**Comment by amirsol81** [comment]
@lukaszgo1 @seanbudd @XLTechie The "Report Destination URL" feature seems to be rather inaccurate or problematic at the moment. I'm using the latest alpha release )27981), and NVDA+K, when pressed on some links, just says "Blank".
For instance:
1. Open the following page: https://www.gsmarena.com/samsung_galaxy_s23_s23_plus_s23_ultra_camera_improvements_march_update-news-58071.php
2. There are some graphical links on the page such as "Samsung Galaxy S23 review", "Samsung Galaxy S23+ review", "Samsung Galaxy S23 series' March update changelog (machine translated", "Our camera shootout video featuring Galaxy S23 Ultra, vivo X90 Pro, Xiaomi 12S Ultra and 13 Pro is out", etc. G and SHIFT+G are also capable of catching them.
3. When NVDA+K is pressed on them, the message "Blank" is announced.
Almost all of them point to other pages, and just one of them refers to the same page. I think the destination should be properly caught and reported.

**Comment by lukaszgo1** [comment]
@amirsol81 I cannot reproduce  this in latest Firefox Nightly. Could you please create a new issue. In the bug report please mention what web browser you';re using, and if this is a regression from this PR, i.e. if this worked when the URL was reported for the navigator object.

**Comment by seanbudd** [comment]
yes, please create an issue. Comments on closed PRs are easily lost.

**Comment by amirsol81** [comment]
@seanbudd Thanks - I will.

**Pull request merged**